### PR TITLE
Added: `function-url-scheme-blacklist` rule.

### PR DESCRIPTION
**Pull request by evilebottnawi**
> Which issue, if any, is this issue related to?

#2513

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

### CODE REVIEW

File: lib/rules/function-url-scheme-blacklist/README.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# function-url-scheme-blacklist
+
+Specify a blacklist of allowed url schemes.
```
Comment (jeddy3) [inline lib/rules/function-url-scheme-blacklist/README.md]: Specify a blacklist of disallowed url schemes.

**Comment by evilebottnawi** [comment]
@jeddy3 stupid error, sorry

File: lib/rules/function-url-scheme-blacklist/index.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+const report = require("../../utils/report")
+const ruleMessages = require("../../utils/ruleMessages")
+const validateOptions = require("../../utils/validateOptions")
+const _ = require("lodash")
+const parse = require("url").parse
+
+const ruleName = "function-url-scheme-blacklist"
+
+const messages = ruleMessages(ruleName, {
+  rejected: scheme => `Unexpected url scheme "${scheme}:"`,
```
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/index.js]: Upper case “URL”.

File: lib/rules/function-url-scheme-blacklist/index.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+})
+
+const rule = function (blacklist) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, {
+      actual: blacklist,
+      possible: [_.isString],
```
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/index.js]: Readme says it accepts `array|string`. I'm not sure if it checks that. There're no tests if only one `string` was in config (not array with one `string`).

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+    code: "a { background: no-repeat center/80% url('https://www.example.com/file.jpg'); }",
+    message: messages.rejected("https"),
+    line: 1,
+    column: 42,
+  } ],
+})
+
+testRule(rule, {
+  ruleName,
+  config: [[]],
```
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: Why this test group rejects `https` and `data`? Is it an undocumented behaviour?

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    code: "a { background-image: url('data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs='); }",
+    message: messages.rejected("data"),
+    line: 1,
+    column: 27,
+  } ],
+})
+
+testRule(rule, {
+  ruleName,
+  config: [""],
```
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: Why this test group rejects `https` and `data`? Is it an undocumented behaviour?

Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: @hudochenkov this used if your want add all schemes to blacklist analogy in whitelist

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js (hunk 4 above)
Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: @hudochenkov this used if your want add all schemes to blacklist analogy in whitelist

File: lib/rules/function-url-scheme-blacklist/index.js (hunk 3 above)
Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/index.js]: @hudochenkov we use this in all `*-blacklist/*-whitelist` rules.

File: lib/rules/function-url-scheme-blacklist/index.js (hunk 2 above)
Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/index.js]: done, also in `whitelist` rule.

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js (hunk 5 above)
Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: @hudochenkov we already have same tests in `function-url-scheme-whitelist`

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js (hunk 4 above)
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: But still behaviour isn't documented anywhere. We should document this.

File: lib/rules/function-url-scheme-blacklist/__tests__/index.js (hunk 5 above)
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: But still behaviour isn't documented anywhere. We should document this.

Comment (evilebottnawi) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: @hudochenkov can your help me with this, I would like the current time to concentrate on the new rules (https://github.com/stylelint/stylelint/issues/2528) and I'm bad at documentation :disappointed:

Comment (jeddy3) [inline lib/rules/function-url-scheme-blacklist/__tests__/index.js]: @hudochenkov Good catch! Sorry I missed this. This isn't the correct behaviour. Both `[""]` and `[[]]` should _allow all_ schemes. 

This is how the other `-blacklist` work. `-whitelist` work differently. 

It's useful to think as

`blacklist: []` = "blacklist nothing" i.e. "everything is ok"
`whitelist: []` = "whitelist nothing" i.e. "nothing is ok".

**Comment by evilebottnawi** [comment]
@jeddy3 can your help with documentation this?

**Comment by jeddy3** [comment]
> @jeddy3 can your help with documentation this?

It's a behaviour issue, not a documentation one. I'm pushing up a commit to fix though.

**Comment by jeddy3** [comment]
- I've added to tests to the `property-*list` rules to show how the whitelist and blacklist rule behaviours are _correctly_ different when given empty strings or arrays as primary options.
- This behaviour comes for free when using the `matchesStringOrRegExp` util. I've updated `function-url-scheme-blacklist` to use it.
- Likewise, support for regex also comes for free when using `matchesStringOrRegExp`.
- I've removed the tests for case insensitive _primary options_. None of the other `*-*list` rules have case-insensitive _primary options_. If we want to revisit this behaviour, so should do so across the board and in a separate issue.

**Comment by evilebottnawi** [comment]
@jeddy3 After commits what are my actions here? Or just wait when @hudochenkov do review?

**Comment by jeddy3** [comment]
> Or just wait when @hudochenkov do review?

Just wait on @hudochenkov as my commits came off the back of his original review.

File: lib/rules/function-url-scheme-blacklist/README.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+```css
+a { background-image: url('http://www.example.com/file.jpg'); }
+/**                        ↑
+ *           This URL scheme */
+```
+
+A [URL scheme](https://url.spec.whatwg.org/#syntax-url-scheme) consists of alphanumeric, `+`, `-`, and `.` characters. It can appear at the start of a URL and is followed by `:`.
+
+This rule treats URL schemes as case insensitive (`https` and `HTTPS` are the same).
```
Comment (hudochenkov) [inline lib/rules/function-url-scheme-blacklist/README.md]: > I've removed the tests for case insensitive primary options. None of the other *-*list rules have case-insensitive primary options. If we want to revisit this behaviour, so should do so across the board and in a separate issue. — @jeddy3 

Is this line in readme still valid?

Comment (jeddy3) [inline lib/rules/function-url-scheme-blacklist/README.md]: Yes. I've added tests to show this.

Like all the other `*-*list` rules, this rule now _only_ accepts lowercase primary options but is case insensitive _when checking for matches_.

So `function-url-scheme-blacklist: "http"` will disallow both `http://g.co/x.jpg` and `HTTP://g.co/x.jpg`.

https://github.com/stylelint/stylelint/issues/2660 is about allowing the user to specify primary options in _any case_ e.g. `function-url-scheme-blacklist: "HTTP"`, or `function-blacklist: "translateX"`.

**Comment by jeddy3** [comment]
> And again about "" and [] configurations. We won't document them?

I'm not sure there's any need to do that now because the behaviour is intuitive. We don't document this behaviour in any of the other `*-blacklist` rules for this reason.

`scheme-blacklist: []` = "blacklist nothing" i.e. "all schemes are ok" seems pretty clear to me.

**Comment by jeddy3** [comment]
Changelog 

> -   Added: `function-url-scheme-blacklist` rule ([#2626](https://github.com/stylelint/stylelint/pull/2626)).

**Pull request merged**